src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs

```diff
@@ -195,66 +195,76 @@ public override void VisitParenthesizedLambdaExpression(ParenthesizedLambdaExpre
 
         public override void VisitLocalFunctionStatement(LocalFunctionStatementSyntax node)
         {
-            var body = (CSharpSyntaxNode)node.Body ?? node.ExpressionBody;
-            LocalFunctionSymbol match = null;
-            // Don't use LookupLocalFunction because it recurses up the tree, as it
-            // should be defined in the directly enclosing block (see note below)
+            bool oldSawYield = _sawYield;
+            Binder binder = _enclosing;
+            Symbol oldMethod = null;
+            LocalFunctionSymbol match = FindLocalFunction(node, _enclosing);
 
-            Binder possibleScopeBinder = _enclosing;
-            while (possibleScopeBinder != null && !possibleScopeBinder.IsLocalFunctionsScopeBinder)
+            if ((object)match != null)
             {
-                possibleScopeBinder = possibleScopeBinder.Next;
+                oldMethod = _containingMemberOrLambda;
+                _containingMemberOrLambda = match;
+
+                binder = match.IsGenericMethod
+                    ? new WithMethodTypeParametersBinder(match, _enclosing)
+                    : _enclosing;
+
+                binder = binder.WithUnsafeRegionIfNecessary(node.Modifiers);
+                binder = new InMethodBinder(match, binder);
             }
 
-            if (possibleScopeBinder != null)
+            BlockSyntax blockBody = node.Body;
+            if (blockBody != null)
             {
-                foreach (var candidate in possibleScopeBinder.LocalFunctions)
+                _sawYield = false;
+                Visit(blockBody, binder);
+
+                if (_sawYield)
                 {
-                    if (candidate.Locations[0] == node.Identifier.GetLocation())
-                    {
-                        match = candidate;
-                    }
+                    _methodsWithYields.Add(blockBody);
                 }
             }
 
-            bool oldSawYield = _sawYield;
-            _sawYield = false;
-
-            if (match != null)
+            ArrowExpressionClauseSyntax arrowBody = node.ExpressionBody;
+            if (arrowBody != null)
             {
-                var oldMethod = _containingMemberOrLambda;
-                _containingMemberOrLambda = match;
+                _sawYield = false;
+                Visit(arrowBody, binder);
+                Debug.Assert(!_sawYield);
+            }
 
-                if (body != null)
-                {
-                    Binder binder = match.IsGenericMethod
-                        ? new WithMethodTypeParametersBinder(match, _enclosing)
-                        : _enclosing;
+            if ((object)oldMethod != null)
```

Comment:
>oldMethod [](start = 24, length = 9)

Minor: It feels like we can store _containingMemberOrLambda and restore it unconditionally.